### ISSUE TITLE
[feature] integrate systemd D-Bus API for unit events

### ISSUE THREAD

**Issue by dustin-decker** [capsule8/capsule8#122]
I'd like to add this feature, but wanted to confer here first to make sure I wasn't stepping other efforts and to get some input.

For now i'd like to focus on these important events:
- unit starts/stops/restarts

Anything else that'd be a priority to check for?

[This](https://godoc.org/github.com/coreos/go-systemd/dbus#Conn.SubscribeUnits) looks like a good way to try and I don't requires bringing in cgo.

**Comment by ddz** [comment]
Hi @dustin-decker,

I think this is a good idea, but has a few challenges with our immediate roadmap. We had previously used the Docker API (long ago) and then `inotify` to monitor container lifecycle events, but found that doing so outside of the `perf` event stream made reliable sequencing difficult (e.g. you get process events inside a container well before the container start event). We just recently transitioned to gathering the same information via kprobes to keep the events in the same sequence as all of the other event types.

Similarly, monitoring systemd unit events via D-Bus would likely have the same challenge, especially once we start surfacing the `cgroup` membership in process events. Are you interested in investigating whether there might be some kprobes that can be used to identify systemd unit lifecycle events?

**Comment by dustin-decker** [comment]
Learning a ton of stuff right now. I think I stumbled on a good way to do this with `kprobes`. 

If you run strace on PID 1 and issue unit changes, you can observe some `sendmsg()` syscalls to the `/run/systemd/private` socket:

`sudo strace -yy -e trace=sendmsg -tfp 1`

(in another terminal `sudo systemctl restart sd.service`)

Some of the messages are obvious lifecycle events that we're interested in:

```
13:25:46 sendmsg(18<UNIX:[256542->2431]>, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="PRIORITY=3\nSYSLOG_FACILITY=3\nCOD"..., iov_len=131}, {iov_base="MESSAGE_ID=be02cf6855d2428ba40df"..., iov_len=43}, {iov_base="\n", iov_len=1}, {iov_base="UNIT=sd.service", iov_len=15}, {iov_base="\n", iov_len=1}, {iov_base="MESSAGE=Failed to start Test sd_"..., iov_len=41}, {iov_base="\n", iov_len=1}, {iov_base="RESULT=failed", iov_len=13}, {iov_base="\n", iov_len=1}], msg_iovlen=9, msg_controllen=0, msg_flags=0}, MSG_NOSIGNAL) = 247
```

I'll look into adding some filters and decoders around those and I think it'll be rolling!

Kind of a bummer that we have to do it at this level, because this socket isn't really an API that the the systemd developers have committed to keeping stable, as far as I know. Luckily docker container configs are probably conforming to OCI spec now that they are on containerd.

**Comment by dustin-decker** [comment]
Yes, it looks like we can probe `sendmsg()` syscalls for D-Bus sockets :D

`sudo strace -yy -e trace=sendmsg -tfp $(sudo pgrep dbus-daemon | head -n 1)`

Trigger some unit changes and observe messages to `/run/dbus/system_bus_socket`. This is great because it might have a more stable API and there is other interesting telemetry D-Bus can provide in the future besides unit lifecycle events.

**Comment by ddz** [comment]
While using the most stable interfaces possible is best, the Sensor currently uses many kprobes on internal Linux kernel functions, which are definitely *not* a stable API.

Have you looked into the existing NetworkEvents, which are accessible through both the gRPC Telemetry API (https://github.com/capsule8/capsule8/blob/master/api/v0/subscription.proto#L238) as well as internal Subscription interfaces? An expression on the `NetworkEventFilter` should be able to restrict events to `sendmsg`s to that socket (and the expression will be evaluated in-kernel, for higher performance).

**Comment by dustin-decker** [comment]
Thanks for the tip. I'm slowly figuring out how to process this, I think.

Here are my steps so far:

1. Find a kernel symbol to attach a kprobe to:

    `sudo cat /sys/kernel/debug/tracing/available_filter_functions | grep sendmsg`

    Looks like I can probably use `SyS_sendmsg` or `sock_sendmsg`.

1. Find out the arguments (requires kernel-debuginfo):

    ```
    $ sudo perf probe -V sock_sendmsg
    Available variables at sock_sendmsg
            @<sock_sendmsg+0>
                    struct msghdr*  msg
                    struct socket*  sock
    ```

    Those should occupy the first two registers, `rdx`, and `rsi` for `x86_64`.
    That makes the implementation architecture-dependent.
    
    Both are pointers that need to be dereferenced, which `kprobe` supports doing.

1. Run kprobe:

    ```
    $ sudo ./bin/kprobe -symbol sock_sendmsg -fetchargs "fd=+0(%di) data=+0(%si)" | head -n 1
    event:<id:"c8be0595dc0c00f9efd94f65726ca7b9ad7000c6542b9c95dd915515608e5342" process_id:"37fd87deafc01bcc823244eac209ec5aee699e2ead3f0fb37253702b5ccbfc5f" process_pid:11375 sensor_id:"2eb42b2f87c6c35b9c227a9ab5bd4df3c99b84a500c47f809ade83246bcbba6f" sensor_sequence_number:724 sensor_monotime_nanos:2246363716094 kernel_call:<arguments:<key:"__probe_ip" value:<field_type:UINT64 unsigned_value:18446744072542691088 > > arguments:<key:"common_flags" value:<field_type:UINT8 unsigned_value:1 > > arguments:<key:"common_pid" value:<field_type:SINT32 signed_value:11375 > > arguments:<key:"common_preempt_count" value:<field_type:UINT8 unsigned_value:0 > > arguments:<key:"common_type" value:<field_type:UINT16 unsigned_value:1924 > > arguments:<key:"data" value:<field_type:UINT64 unsigned_value:0 > > arguments:<key:"fd" value:<field_type:UINT64 unsigned_value:18446462603027808259 > > > credentials:<uid:1000 gid:1000 euid:1000 egid:1000 suid:1000 sgid:1000 fsuid:1000 fsgid:1000 > process_tgid:11334 > 
    ```

    The data is not readable right now.. I think we need to extract it and cast it from various offsets in the register.

I assume my next step is to figure out offsets and types of the data structure like you guys have done [here](https://github.com/capsule8/capsule8/blob/03b002ffa7d17d74799234604a5fae01f0d6563c/pkg/sensor/network.go#L37).
I'd figure out that somehow from the [data structures](https://github.com/systemd/systemd/blob/7fe2903c2315db9d9f501ae255a6b6e4f01ba46c/src/libsystemd/sd-bus/bus-socket.c#L1044) that systemd writes to the DBUS socket?

Another potential option could be hooking functions in systemd with `uprobes`, but that seems more fragile.

Any tips to go from here? This is new territory for me.

**Comment by dustin-decker** [comment]
I might be on the wrong track because those `sendmsg` syscalls happen after the `exec` syscall, so they'd already be lagging in the `perf` event stream...

I don't systemd places convenient JSON files like containerd does before starting (not documented that I saw, at least), so I think doing this with the `perf` tooling is going to be non-trivial for my current skill level.
